File: src/2025-03-04-12-20-21--deepseek-vr-0304/cpp/exercises/practice/bank-account/bank_account.cpp
#include "bank_account.h"

namespace Bankaccount {

Bankaccount::Bankaccount() : balance_(0), is_closed(true) {}

void Bankaccount::open() {
    std::lock_guard<std::mutex> lock(mtx);
    is_closed = false;
    balance_ = 0;
}

void Bankaccount::deposit(int amount) {
    std::lock_guard<std::mutex> lock(mtx);
    if (is_closed) {
        throw std::runtime_error("Account closed");
    }
    if (amount <= 0) {
        throw std::runtime_error("Invalid amount");
    }
    balance_ += amount;
}

void Bankaccount::withdraw(int amount) {
    std::lock_guard<std::mutex> lock(mtx);
    if (is_closed) {
        throw std::runtime_error("Account closed");
    }
    if (amount <= 0) {
        throw std::runtime_error("Invalid amount");
    }
    if (balance_ < amount) {
        throw std::runtime_error("Insufficient funds");
    }
    balance_ -= amount;
}

int Bankaccount::balance() const {
    std::lock_guard<std::mutex> lock(mtx);
    if (is_closed) {
        throw std::runtime_error("Account closed");
    }
    return balance_;
}

void Bankaccount::close() {
    std::lock_guard<std::mutex> lock(mtx);
    is_closed = true;
}

}  // namespace Bankaccount

File: src/2025-03-04-12-20-21--deepseek-vr-0304/java/exercises/practice/bank-account/src/main/java/BankAccount.java
class BankAccount {
    private boolean isOpen;
    private int balance;

    synchronized void open() throws BankAccountActionInvalidException {
        if (isOpen) {
            throw new BankAccountActionInvalidException("Account is already open");
        }
        isOpen = true;
        balance = 0;
    }

    synchronized void close() throws BankAccountActionInvalidException {
        if (!isOpen) {
            throw new BankAccountActionInvalidException("Account is not open");
        }
        isOpen = false;
    }

    synchronized int getBalance() throws BankAccountActionInvalidException {
        if (!isOpen) {
            throw new BankAccountActionInvalidException("Account is closed");
        }
        return balance;
    }

    synchronized void deposit(int amount) throws BankAccountActionInvalidException {
        if (!isOpen) {
            throw new BankAccountActionInvalidException("Account is closed");
        }
        if (amount <= 0) {
            throw new BankAccountActionInvalidException("Cannot deposit or withdraw non-positive amount");
        }
        balance += amount;
    }

    synchronized void withdraw(int amount) throws BankAccountActionInvalidException {
        if (!isOpen) {
            throw new BankAccountActionInvalidException("Account is closed");
        }
        if (amount <= 0) {
            throw new BankAccountActionInvalidException("Cannot deposit or withdraw non-positive amount");
        }
        if (balance < amount) {
            throw new BankAccountActionInvalidException("Cannot withdraw more than balance");
        }
        balance -= amount;
    }

}
